### ISSUE TITLE
Duplicate UDP socket bind not generating an error

### ISSUE THREAD

**Issue by sipsorcery** [dotnet/runtime#36618]
### Description

I had a sporadic issue where I was successfully binding to a UDP socket but then did not receive any incoming packets. I used a packet trace to confirm the packets were being sent and the receive code was not much more than BeginReceiveFrom/EndReceiveFrom.

Eventually I spotted that the system generated port allocated for my bind was a being assigned to two different processes so in effect there ended up being two UDP sockets listening on the same port with only one being able to receive packets.

The issue occurs if binds to the same port are done using `IPAddress.Any` and a dual mode socket with `IPAddress.IPv6Any`. The second process attempting the bind on the duplicate port succeeds BUT will not receive any packets until the first socket closes. The two binds can be in different processes and both will end up thinking they have a good socket whereas only the one that bound first will be capable of receiving.

### Configuration

Windows 10 Pro 10.0.19041
.Net Core 3.1.202

### Regression?

I don't believe so.

### Other information

The program below demonstrates the problem.

```c#
using System;
using System.Net;
using System.Net.Sockets;

namespace UdpDuplicateBindTest
{
    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("UDP Duplicate Bind Test");

            var socketAny = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
            socketAny.Bind(new IPEndPoint(IPAddress.Any, 0));
            IPEndPoint anyEP = socketAny.LocalEndPoint as IPEndPoint;

            Console.WriteLine($"Socket successfully bound on {anyEP}");

            var socketIP6Any = new Socket(AddressFamily.InterNetworkV6, SocketType.Dgram, ProtocolType.Udp);
            socketIP6Any.DualMode = true;
            socketIP6Any.Bind(new IPEndPoint(IPAddress.IPv6Any, anyEP.Port));
            IPEndPoint ip6AnyEP = socketIP6Any.LocalEndPoint as IPEndPoint;

            Console.WriteLine($"Socket successfully bound on {ip6AnyEP}, is dual mode {socketIP6Any.DualMode}");

            Console.ReadLine();
        }
    }
}
```

Output:

````
UDP Duplicate Bind Test
Socket successfully bound on 0.0.0.0:61696
Socket successfully bound on [::]:61696, is dual mode True
````

**Comment by sipsorcery** [comment]
Additional notes:

 - The problem can be circumvented on Windows by setting the exclusive use socket option (it was my experience up until now that exclusive use behaviour was default):
`sock.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse, true)`

 - The problem does not exist on Ubuntu. The duplicate socket bind throws a `SocketException`:
````
  Error Message:
   System.Net.Sockets.SocketException : Address already in use
  Stack Trace:
     at System.Net.Sockets.Socket.UpdateStatusAfterSocketErrorAndThrowException(SocketError error, String callerName)
   at System.Net.Sockets.Socket.DoBind(EndPoint endPointSnapshot, SocketAddress socketAddress)
   at System.Net.Sockets.Socket.Bind(EndPoint localEP)
````

**Comment by antonfirsov** [comment]
### TLDR
The issue and the difference between Linux and Windows is rooted in the default behavior of sockets created by the OS-es. The related socket options are currently untouched during .NET socket creation.

### Details

**On Windows** and in the .NET `Socket` API (which is modeled on Windows behavior) there are two socket options controlling exclusivity of socket addresses: `SO_REUSEADDR` (`SocketOptionName.ReuseAddress`) and `SO_EXCLUSIVEADDRUSE` (`SocketOptionName.ExclusiveAddressUse`). These options define a complicated behavior where conflicts between `Specific <-> Specific`, `Specific <-> Wildcard` addresses [are handled differently](https://docs.microsoft.com/en-us/windows/win32/winsock/using-so-reuseaddr-and-so-exclusiveaddruse#using-so_exclusiveaddruse) for different configurations. The default values are `SO_REUSEADDR == 0` and `SO_EXCLUSIVEADDRUSE == 0`.

With the these defaults, the IPV4 endpoint `0.0.0.0:4242` does not conflict with endpoint `[::]:4242`, since these are recognized to be different (and `SO_EXCLUSIVEADDRUSE` is not enabled to exclude conflicts with wildcard addresses). However the IPV4 socket address `0.0.0.0:4242` will do conflict with the IPV4-mapped dual-stack socket address `::ffff:0.0.0.0/4242`, since those are recognized to be the same by the OS.

**On Unix** there is no such complicated mechanism, both `SocketOptionName.ReuseAddress` and `SocketOptionName.ExclusiveAddressUse` are mapped to the same native flag `SO_REUSEPORT` as following: `SocketOptionName.ReuseAddress == SO_REUSEPORT` and `SocketOptionName.ExclusiveAddressUse == !SO_REUSEPORT`

The default value of `SO_REUSEPORT` is `0`, which makes `SocketOptionName.ReuseAddress` default to `1`, and disables the reusing of socket addresses.
*(Note: the Unix PAL also makes sure that `SO_REUSEADDR == SO_REUSEPORT` by setting those values together.)

Another related (and weird) thing I experienced on Linux, that `SO_REUSEPORT == 0` while works well with UDP, actually doesn't disallow binding TCP sockets to the same address. So far this is full magic for me. (Similar to #30782 probably)

### TLDR2: conclusions and open questions
- The behavior is quite inconsistent between operating systems, which might be unexpected to users
- Fixing the inconsistency would mean breaking exisitng behavior (Changing default value of `ExclusiveAddressUse` on either Unix or Windows)
- Do we want to risk a breaking change?
- This is not the first time where an OS default is leading to confusion. In general: what has higher priority? OS compatibility or cross-platform consistence? I would be happy to see a design guideline here.
- The issue might be also addressed by improving documentation of [`SocketOptionName`](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socketoptionname)

/cc @wfurt @scalablecory @tmds

**Comment by karelz** [comment]
At minimum, this does not seem to be critical for 5.0. If we are exposing the native behavior of Sockets and there is no clear "right way" to do it (at least not now without deeper analysis), we should keep it without breaking now (it is not a regression), until we have super-clear plan what to do in the space -- and then do intentional breaking change when we're ready.

**Comment by tmds** [comment]
Imo the preferable default is to be 'exclusive', which means other applications cannot unexpectedly steal ports. That is how it is on Unix. Though for backwards compatibility, I would keep the Windows behavior.

An application that cares enough should set the `Socket.ExclusiveAddressUse` property to `true` (to prevent stealing) or `SocketOptionName.ReuseAddress` to `true` (to allow stealing).
